File: solutions/arrays-hashing/448-find-all-numbers-disappeared-in-an-array.py
"""
# 448. Find All Numbers Disappeared In An Array
# Difficulty: Medium
Given a problem that demonstrates key concepts in Arrays Hashing.

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
[This problem requires understanding of arrays hashing concepts. The key insight is to identify the optimal approach for this specific scenario.]

### APPROACH:
1. **Analyze the problem**: Understand the input constraints and expected output
2. **Choose the right technique**: Apply arrays hashing methodology
3. **Implement efficiently**: Focus on optimal time and space complexity
4. **Handle edge cases**: Consider boundary conditions and special cases

### WHY THIS WORKS:
- The solution leverages arrays hashing principles
- Time complexity is optimized for the given constraints
- Space complexity is minimized where possible

### TIME COMPLEXITY: O(n)
### SPACE COMPLEXITY: O(1)

### EXAMPLE WALKTHROUGH:
```
Input: [example input]
Step 1: [explain first step]
Step 2: [explain second step]
Output: [expected output]
```

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>

<details>
<summary><b>💡 APPROACH</b></summary>

The approach uses arrays hashing techniques to solve this problem efficiently.

### Algorithm Steps:
1. Initialize necessary variables
2. Process input using arrays hashing method
3. Return the computed result

</details>
"""

class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        """
        Find all numbers from 1 to n that are missing in the input array.
        
        Args:
            nums: List of integers where each integer is in range [1, n]
                 and n is the length of the list
        
        Returns:
            List of integers that are missing from the input array
        
        Time Complexity: O(n)
        Space Complexity: O(1) - excluding the output array
        """
        # Mark present numbers using index manipulation
        for num in nums:
            # Get the absolute value since numbers might have been marked negative
            index = abs(num) - 1
            # Mark as seen by making the number at index negative
            nums[index] = -abs(nums[index])
        
        # Find missing numbers by checking which indices contain positive numbers
        result = []
        for i in range(len(nums)):
            if nums[i] > 0:
                # i + 1 is missing from the original array
                result.append(i + 1)
        
        return result

def test_solution():
    """
    Test cases for 448. Find All Numbers Disappeared In An Array.
    """
    solution = Solution()

    # Test case 1: Basic functionality
    # result = solution.solve([test_input])
    # expected = [expected_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Edge case
    # result = solution.solve([edge_case_input])
    # expected = [edge_case_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 448. Find All Numbers Disappeared In An Array")

File: solutions/arrays-hashing/448-find-all-numbers-disappeared-in-an-array.js
I'll help you create a solution for finding disappeared numbers in JavaScript. I'll write the code with clear comments and proper structure.

```javascript
/**
 * Find Disappeared Numbers
 * 
 * Problem: Given an array nums of n integers where nums[i] is in the range [1, n],
 * return an array of all integers in the range [1, n] that do not appear in nums.
 * 
 * @param {number[]} nums - Input array containing integers
 * @return {number[]} - Array of numbers that are missing from the input array
 */

/**
 * Main function to find disappeared numbers
 * Time Complexity: O(n)
 * Space Complexity: O(1) - excluding the output array
 */
function findDisappearedNumbers(nums) {
    // Edge case: empty array
    if (!nums || nums.length === 0) {
        return [];
    }

    const n = nums.length;
    const result = [];

    // Mark numbers that are present by making their corresponding indices negative
    for (let i = 0; i < n; i++) {
        // Get the absolute value since the number might have been marked negative
        const num = Math.abs(nums[i]);
        // Mark the number as seen by making the value at index (num-1) negative
        if (nums[num - 1] > 0) {
            nums[num - 1] *= -1;
        }
    }

    // Find indices that are still positive (these represent missing numbers)
    for (let i = 0; i < n; i++) {
        if (nums[i] > 0) {
            // Add 1 to index to get the actual missing number
            result.push(i + 1);
        }
    }

    return result;
}

// Example usage and test cases
function runTests() {
    const testCases = [
        [4,3,2,7,8,2,3,1],  // Expected: [5,6]
        [1,1],              // Expected: [2]
        [1,2,3,4],          // Expected: []
        [],                 // Expected: []
        [1,1,1,1],          // Expected: [2,3,4]
    ];

    testCases.forEach((test, index) => {
        const result = findDisappearedNumbers(test);
        console.log(`Test ${index + 1}:`);
        console.log(`Input: [${test}]`);
        console.log(`Output: [${result}]`);
        console.log('---');
    });
}

// Export the function for use in other modules
module.exports = findDisappearedNumbers;

// Run tests if this file is being run directly
if (require.main === module) {
    runTests();
}
```

This implementation provides a solution to the "Find Disappeared Numbers" problem with the following features:

1. **Efficient Algorithm**:
   - Uses O(n) time complexity
   - Uses O(1) space complexity (excluding the output array)
   - Modifies the input array to mark presence of numbers

2. **Clean Code Structure**:
   - Clear function and variable names
   - Proper comments explaining the logic
   - Modular design with separate test function

3. **Error Handling**:
   - Handles edge cases like empty arrays
   - Handles arrays with duplicate numbers

4. **Testing**:
   - Includes test cases covering various scenarios
   - Easy to run and verify functionality

5. **Module Export**:
   - Can be imported and used in other files
   - Includes standalone test execution

The algorithm works by:
1. Using the input array itself as a hash table
2. Marking presence of numbers by making corresponding indices negative
3. Finding positive numbers in the final array to identify missing numbers

This solution follows JavaScript best practices and provides a clean, efficient implementation of the problem.